app/todos/upsert/upsert-todo.controller.js

```diff
@@ -3,7 +3,119 @@
 angular.module('todoApp.todos')
     .controller('UpsertTodoController', [
         '$scope',
-        function($scope) {
+        '$state',
+        '$stateParams',
+        '$http',
+        '$q',
+        function ($scope,
+                  $state,
+                  $stateParams,
+                  $http,
+                  $q) {
+
+            var vm = this;
+            vm.tagsLoading = true;
+            vm.loading = false;
+            vm.checkboxesTouched = false;
+            vm.isNew = ($stateParams.id === '');
+            vm.todo = null;
+            vm.tags = [];
+
+
+            vm.someTagsSelected = someTagsSelected;
+            vm.upsertTodo = upsertTodo;
+
+            init();
+
+            function init() {
+                vm.tagsLoading = true;
+                $q.all([getTodo(), getTags()])
+                    .then(function (responses) {
+                        vm.todo = responses[0].data;
+                        vm.tags = mapTags(responses[1].data.tags);
+                        console.log(vm.todo);
+                        vm.tagsLoading = false;
+                    })
+                    .catch(function (error) {
+                        console.log('Error loading tags/todo');
+                        vm.tagsLoading = false;
+                    });
```

Comment:
In the above promise, a parallel request if made to get all tags and optionally to get todo in case it is an **Edit Todo** form.